### PR TITLE
[FLINK-28143][runtime] Introduce BlocklistHandler

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change

Introduce BlocklistHandler

## Verifying this change

Unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

**Review by zhuzhurk — COMMENTED** [review]
Thanks for opening this PR! @wanglijie95 
I have a few comments. Please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlockedNode.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    public static BlockedNode merge(BlockedNode node, BlockedNode other) {
+        checkNotNull(node);
+        checkNotNull(other);
+        checkState(node.nodeId.equals(other.nodeId));
+
+        Set<String> mergedCauses = new HashSet<>();
+        mergedCauses.addAll(node.causes);
+        mergedCauses.addAll(other.causes);
+
+        long mergedEndTimestamp = Math.max(node.endTimestamp, other.endTimestamp);
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlockedNode.java:70]: Make sense. Assume the existing one is A, new one is B, A.endTimestamp == B.endTimestamp and A.cause != B.cause. Then I think the merged one should be [B.endTimestamp, B.cause] . WDYT?

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlockedNode.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** This class represents a blocked node record. */
+public class BlockedNode implements Serializable {
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlockedNode.java:31]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/DefaultBlocklistTracker.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    private Optional<BlockedNode> tryAddOrMerge(BlockedNode newNode) {
+        checkNotNull(newNode);
+        final String nodeId = newNode.getNodeId();
+        final BlockedNode existingNode = blockedNodes.get(nodeId);
+
+        if (existingNode == null) {
+            blockedNodes.put(nodeId, newNode);
+            return Optional.of(newNode);
+        } else if (!existingNode.equals(newNode)) {
+            BlockedNode merged = BlockedNode.merge(existingNode, newNode);
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/DefaultBlocklistTracker.java:51]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistTracker.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+     * @return true if the given node is blocked, otherwise false
+     */
+    boolean isBlockedNode(String nodeId);
+
+    /**
+     * Get all blocked node ids.
+     *
+     * @return a collection containing all blocked node ids
+     */
+    Collection<String> getAllBlockedNodeIds();
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistTracker.java:48]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistContext.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+     * @param blockedNodes the blocked nodes
+     */
+    void blockResources(Collection<BlockedNode> blockedNodes);
+
+    /**
+     * Unblock resources on the nodes.
+     *
+     * @param unBlockedNodes the unBlocked nodes
+     */
+    void unBlockResources(Collection<BlockedNode> unBlockedNodes);
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistContext.java:38]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistContext.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.util.Collection;
+
+/** This class is responsible for blocking and unblocking resources. */
+public interface BlocklistContext {
+
+    /**
+     * Block resources on the nodes.
+     *
+     * @param blockedNodes the blocked nodes
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistContext.java:29]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistContext.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+     */
+    void blockResources(Collection<BlockedNode> blockedNodes);
+
+    /**
+     * Unblock resources on the nodes.
+     *
+     * @param unBlockedNodes the unBlocked nodes
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistContext.java:36]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistHandler.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+public interface BlocklistHandler {
+
+    /**
+     * Block a node.
+     *
+     * @param nodeId ID of the node
+     * @param cause the cause for blocking the node
+     * @param endTimestamp the timestamp at which the node should be unblocked
+     */
+    default void blockNode(String nodeId, String cause, long endTimestamp) {
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlocklistHandler.java:38]: OK

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/DefaultBlocklistHandler.java [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+                    listener -> listener.notifyNewBlockedNodes(newlyAddedOrMerged));
+            blocklistContext.blockResources(newlyAddedOrMerged);
+        }
+    }
+
+    @Override
+    public boolean isBlockedTaskManager(ResourceID taskManagerId) {
+        assertRunningInMainThread();
+
+        return taskManagerToNode.containsKey(taskManagerId)
```
[the remark this answers is not in the document]
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/DefaultBlocklistHandler.java:142]: You are right, we never expect it fails to retrieve the node id of a task manager.

**Comment by wanglijie95** [comment]
Thanks for review @zhuzhurk. I 've addressed all comments, please take a look.

File: flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlockedNode.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zhuzhurk) [reply flink-runtime/src/main/java/org/apache/flink/runtime/blocklist/BlockedNode.java]: Looks good.

**Review by zhuzhurk — APPROVED** [review]
Thanks for addressing all the comments. @wanglijie95 
The change looks good to me.
Will squash it and merge it once CI is green.

**Pull request closed without merging**